source/core/shape.cc

```diff
@@ -484,17 +487,166 @@ double
+      nodes[n_i][0] = shape_arguments[3 * number_of_nodes + n_i];
+      nodes[n_i][1] = shape_arguments[4 * number_of_nodes + n_i];
+      nodes[n_i][2] = shape_arguments[5 * number_of_nodes + n_i];
+    }
+  initialize_bounding_box();
+}
+
+template <int dim>
+double
+RBFShape<dim>::value(const Point<dim> &evaluation_point,
+                     const unsigned int /*component*/) const
+{
+  Point<dim> centered_point = this->align_and_center(evaluation_point);
+
+  double bounding_box_distance = bounding_box->value(centered_point);
+  double value                 = std::max(bounding_box_distance, 0.0);
+
+  double dist, basis;
+  size_t number_of_nodes = weights.size();
+  // Algorithm inspired by Optimad Bitpit. https://github.com/optimad/bitpit
+  for (size_t i = 0; i < number_of_nodes; ++i)
+    {
+      dist  = (centered_point - nodes[i]).norm() / support_radii[i];
+      basis = evaluate_basis_function(basis_functions[i], dist);
+      value += basis * weights[i];
+    }
+  return value;
+}
+
+template <int dim>
+std::shared_ptr<Shape<dim>>
+RBFShape<dim>::static_copy() const
+{
+  std::shared_ptr<Shape<dim>> copy =
+    std::make_shared<RBFShape<dim>>(this->support_radii,
+                                    this->basis_functions,
+                                    this->weights,
+                                    this->nodes,
+                                    this->position,
+                                    this->orientation);
+  return copy;
+}
+
+template <int dim>
+double
+RBFShape<dim>::displaced_volume(const double fluid_density)
+{
+  return bounding_box->displaced_volume(fluid_density);
+}
+
+template <int dim>
+void
+RBFShape<dim>::initialize_bounding_box()
+{
+  Point<dim>     high_bounding_point{};
+  Point<dim>     low_bounding_point{};
+  Point<dim>     bounding_box_center{};
+  Tensor<1, dim> half_lengths = Tensor<1, dim>();
+  for (int d = 0; d < dim; d++)
+    {
+      high_bounding_point[d] = std::numeric_limits<double>::lowest();
+      low_bounding_point[d]  = std::numeric_limits<double>::max();
+      for (size_t i = 0; i < number_of_nodes; i++)
+        {
+          if (low_bounding_point[d] > nodes[i][d])
+            low_bounding_point[d] = nodes[i][d];
+          if (high_bounding_point[d] < nodes[i][d])
+            high_bounding_point[d] = nodes[i][d];
+        }
+      bounding_box_center[d] =
+        0.5 * (low_bounding_point[d] + high_bounding_point[d]);
+      half_lengths[d] = 0.5 * (high_bounding_point[d] - low_bounding_point[d]);
+    }
+  bounding_box = std::make_shared<Rectangle<dim>>(half_lengths,
+                                                  bounding_box_center,
+                                                  Tensor<1, 3>());
+}
+
+template <int dim>
+double
+RBFShape<dim>::evaluate_basis_function(const RBFBasisFunction basis_function,
+                                       const double           distance) const
+{
+  switch (basis_function)
```

Comment:
If you ever have only one basis function type for your entire object (if it becomes a requirement of your rbf geometry), this switch case could be removed for optimization. :smile:

Reply:
Good suggestion, thanks!